runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10261,61 +10259,49 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
       //   integer arithmetic, checking carry bit is enough to detect it.
       //   For variable length array, we did an up-front check already.
 
-      static char * disableInitClear = feGetEnv("TR_disableInitClear");
-      static char * disableBatchClear = feGetEnv("TR_DisableBatchClear");
+      // Zero initialize newly allocated array or object if batch clearing is disabled and node can not skip zero initialization.
+      static bool disableBatchClear = feGetEnv("TR_DisableBatchClear") != NULL;
+      bool inlineZeroInitialization = disableBatchClear && !node->canSkipZeroInitialization();
 
-      TR::Register * addressReg = NULL, * lengthReg = NULL, * shiftReg = NULL;
-      if (disableBatchClear && disableInitClear==NULL)
+      if (!isVariableLen)
          {
-         addressReg = cg->allocateRegister();
-         lengthReg = cg->allocateRegister();
-         shiftReg = cg->allocateRegister();
+         if (comp->target().is64Bit())
+            iCursor = genLoadLongConstant(cg, node, allocSize, sizeReg, iCursor, conditions);
+         else
+            iCursor = generateLoad32BitConstant(cg, node, allocSize, sizeReg, true, iCursor, conditions);
+         }
 
+      TR::Register * lengthReg = NULL;
+      if (inlineZeroInitialization && isVariableLen)
+         {
+         lengthReg = cg->allocateRegister();
          if (conditions != NULL)
             {
             conditions->resetIsUsed();
-            conditions->addPostCondition(addressReg, TR::RealRegister::AssignAny);
-            conditions->addPostCondition(shiftReg, TR::RealRegister::AssignAny);
             conditions->addPostCondition(lengthReg, TR::RealRegister::AssignAny);
             }
-         }
-
-      if (isVariableLen)
-         {
-         if (disableBatchClear && disableInitClear==NULL)
-            iCursor = generateRRInstruction(cg, TR::InstOpCode::LGR, node, lengthReg, sizeReg, iCursor);
-         if (!comp->getOption(TR_DisableDualTLH) && node->canSkipZeroInitialization())
+         // Saving size - 1 since XC instruction is used to zero memory.
+         // The size is a positive number therefore the result value can not be negative.
+         if (comp->target().cpu.isAtLeast(OMR_PROCESSOR_S390_Z196))
             {
-            iCursor = generateRXInstruction(cg, TR::InstOpCode::getAddOpCode(), node, sizeReg,
-                  generateS390MemoryReference(metaReg, offsetof(J9VMThread, nonZeroHeapAlloc), cg), iCursor);
+            iCursor = generateRIEInstruction(cg, TR::InstOpCode::getAddHalfWordImmDistinctOperandOpCode(), node, lengthReg, sizeReg, -1);
```

Comment:
If we fail the heap top test, we would branch to OOL - in that case, length register will not be used - Can we move this code after the conditional branch to OOL ?

Reply:
In the heap top test, it adds `heapAlloc` to the size register and compare with `heapTop`. If I move it after OOL, I would need to do `length = size - heapTopRegister - 1` instead of `length = size -1` which requires more instructions.